client/service/src/client/call_executor.rs

```diff
@@ -81,15 +83,23 @@ where
 		Block: BlockT,
 		B: backend::Backend<Block>,
 	{
-		let code = self.wasm_override
+		let code = if let Some(d) = self.wasm_override
 			.as_ref()
 			.map::<sp_blockchain::Result<Option<RuntimeCode>>, _>(|o| {
 				let spec = self.runtime_version(id)?.spec_version;
 				Ok(o.get(&spec, onchain_code.heap_pages))
 			})
 			.transpose()?
-			.flatten()
-			.unwrap_or(onchain_code);
+			.flatten() {
```

Comment:
I find this piece of code unnecessary complicated to read,

I would prefer to do it like this:

```rust
		let spec = self.runtime_version(id)?.spec_version;
		let wasm_override_code = self.wasm_override
			.as_ref()
			.and_then(|o| o.get(&spec, onchain_code.heap_pages));

		if let Some(code) = wasm_override_code = {
			debug!(target: "wasm_overrides", "using WASM override for block {}", id);
			Ok(code)
		} else {
			debug!(target: "wasm_overrides",
				"No WASM override available for block {}, using onchain code",
				id
			);
			Ok(onchain_code)
		}

``

Reply:
iirc `runtime_version` is in the `map` because calls to `runtime_version` can be expensive (requires executing a function in the WASM, and getting some values from the database). So to reduce unnecessary checking it was called conditionally if the overrides were present.

This is definitely more readable though; maybe the extra cost is worth it

Reply:
ah, I see makes sense.